crates/forge-runner/src/lib.rs

```diff
@@ -0,0 +1,437 @@
+            sierra_program,
+            test_cases,
+        }
+    }
+}
+
+pub trait TestCaseFilter {
+    fn should_be_run(&self, test_case: &TestCaseRunnable) -> bool;
+}
+
+#[non_exhaustive]
+pub enum TestCrateRunResult {
+    Ok(TestCrateSummary),
+    Interrupted(TestCrateSummary),
+}
+
+pub async fn run_tests_from_crate(
+    tests: Arc<TestCrate>,
+    runner_config: Arc<RunnerConfig>,
+    runner_params: Arc<RunnerParams>,
+    tests_filter: &impl TestCaseFilter,
+) -> Result<TestCrateRunResult> {
+    let runner = Arc::new(
+        SierraCasmRunner::new(
+            tests.sierra_program.clone(),
+            Some(MetadataComputationConfig::default()),
+            OrderedHashMap::default(),
+        )
+        .context("Failed setting up runner.")?,
+    );
+
+    let mut tasks = FuturesUnordered::new();
+    let test_cases = &tests.test_cases;
+    // Initiate two channels to manage the `--exit-first` flag.
+    // Owing to `cheatnet` fork's utilization of its own Tokio runtime for RPC requests,
+    // test execution must occur within a `tokio::spawn_blocking`.
+    // As `spawn_blocking` can't be prematurely cancelled (refer: https://dtantsur.github.io/rust-openstack/tokio/task/fn.spawn_blocking.html),
+    // a channel is used to signal the task that test processing is no longer necessary.
+    let (send, mut rec) = channel(1);
+
+    // The second channel serves as a hold point to ensure all tasks complete
+    // their shutdown procedures before moving forward (more info: https://tokio.rs/tokio/topics/shutdown)
+    let (send_shut_down, mut rec_shut_down) = channel(1);
+
+    for case in test_cases {
+        let case_name = case.name.clone();
+
+        if !tests_filter.should_be_run(case) {
+            tasks.push(tokio::task::spawn(async {
+                Ok(TestCaseSummary::Ignored { name: case_name })
+            }));
+            continue;
+        };
+
+        let function = runner.find_function(&case_name)?;
+        let args = function_args(function, &BUILTINS);
+
+        let case = Arc::new(case.clone());
+        let args: Vec<ConcreteTypeId> = args.into_iter().cloned().collect();
+        let runner = runner.clone();
+
+        tasks.push(choose_test_strategy_and_run(
+            args,
+            case.clone(),
+            runner,
+            runner_config.clone(),
+            runner_params.clone(),
+            &send,
+            &send_shut_down,
+        ));
+    }
+
+    let mut results = vec![];
+    let mut interrupted = false;
+
+    while let Some(task) = tasks.next().await {
+        let result = task??;
+
+        print_test_result(&result);
+
+        if let TestCaseSummary::Failed { .. } = result {
+            if runner_config.exit_first {
+                interrupted = true;
+                rec.close();
+            }
+        }
+
+        results.push(result);
+    }
+
+    // Waiting for things to finish shutting down
+    drop(send_shut_down);
+    let _ = rec_shut_down.recv().await;
```

Comment:
I think this logic is not required anymore. We wait for all tasks in prev loop - I will create ticket to remove and test it